### ISSUE TITLE
Add Attachment support

### ISSUE THREAD

**Issue by bruno-garcia** [getsentry/sentry-react-native#1327]
The native SDKs (sentry-android, sentry-cocoa) added support to Attachments.

The [spec is in the develop docs](https://develop.sentry.dev/sdk/features/#attachments).

Thumbs up if this is the feature you've been waiting for.

**Comment by renchap** [comment]
In addition to this, it would be awesome to be able to run a callback when an error happen and generate attachments from this callback.

I have multiple states in the app (Redux, async storage…) and I would like to dump them in attachments when an issue occurs in the app to be able to understand the error if it comes from a bad state.
Currently I could add them to the event, but it can quickly go over the 200kb limit.

**Comment by marandaneto** [comment]
This is blocked by https://github.com/getsentry/sentry-javascript/issues/4240

The RN SDK uses the JS SDK for building an event and the current version does not support the Envelope feature which is required for sending attachments.

To make it work we'd need to override many of the JS behavior and later replace them once the Envelope feature is landed on the linked PR natively on the JS SDK.

The best is to postpone and implement once #4240 is done, also considering attaching attachments to hard crashes on the Native layer.

Points to consider:

- Use the envelope feature from #4240
- Attach attachments on hard crashes, that requires changes on Sentry-Android, Sentry-Native, and Sentry-Cocoa since we don't sync down the attachments to the Native layers (e.g. KSCrash and Sentry-Native)
- We could technically support an array of bytes and file paths since RN does not support files natively, people most of the time use 3rd party plugins to do so.

**Comment by mmerickel** [comment]
We got hit by lack of support for this feature and the addGlobalEventProcessor approach that's documented for the javascript SDK was not working for us. We'd see 201 Accepted requests for the attachment on the server but they did not get attached to the event. Ultimately we ended up writing a bridge module and using the sentry-cocoa and sentry-android native APIs to send events with attachments for the limited case where we needed an attachment. Would really love support for attachments to get added to sentry-react-native using the `scope.addAttachment` features available in the native sdks.

**Comment by hdsenevi** [comment]
> We got hit by lack of support for this feature and the addGlobalEventProcessor approach that's documented for the javascript SDK was not working for us. We'd see 201 Accepted requests for the attachment on the server but they did not get attached to the event. Ultimately we ended up writing a bridge module and using the sentry-cocoa and sentry-android native APIs to send events with attachments for the limited case where we needed an attachment. Would really love support for attachments to get added to sentry-react-native using the `scope.addAttachment` features available in the native sdks.

Would you consider making the source for this bridge module publicly available? I'm trying to do the same and hit some roadblocks with regards to bridging functionality.

**Comment by renchap** [comment]
#4240 is done, and sentry-react-native 4.0.0 uses it 👀

**Comment by renchap** [comment]
And 4.0.0 is released, maybe you can remove the "blocked" tag? ;)

**Comment by Noitidart** [comment]
@renchap is attachment support in 4.0.0? or 4.0.0 just makes attachment support possible?

**Comment by renchap** [comment]
It should make it possible, but not implemented yet.

**Comment by bruno-garcia** [comment]
JS 7.0 unblocked this and it's a p1 for us. Btw we're hiring a RN engineer if anyone wants to get paid to work on open source full time: https://sentry.io/careers/3642918/

**Comment by marandaneto** [comment]
https://github.com/getsentry/sentry-react-native/pull/2463
